src/qt/bitcoin.cpp

```diff
@@ -352,6 +352,17 @@ void BitcoinApplication::requestShutdown()
     window->setClientModel(nullptr);
     pollShutdownTimer->stop();
 
+#ifdef ENABLE_WALLET
+    // Delete wallet controller here manually, instead of relying on Qt object
+    // tracking (https://doc.qt.io/qt-5/objecttrees.html). This makes sure
+    // walletmodel m_handle_* notification handlers are deleted before wallets
+    // are unloaded, which can simplify wallet implementations. It also avoids
```

Comment:
> Delete wallet controller here manually... This makes sure walletmodel `m_handle_*` notification handlers are deleted before wallets are unloaded, which can simplify wallet implementations.

Sorry, but I cannot get it.
To achieve the goal--notification handler deletion--the `~WalletModel` destructor must be called. But I cannot see the way how deletion of the `m_wallet_controller` could works here, as an instance of the `WalletController` class does _not_ own `WalletModel` objects, because `WalletController::m_wallets` is `std::vector<WalletModel*>`, not `std::vector<std::unique_ptr<WalletModel>>`.

Did I miss the point?

Reply:
> > Delete wallet controller here manually... This makes sure walletmodel `m_handle_*` notification handlers are deleted before wallets are unloaded, which can simplify wallet implementations.
> 
> Sorry, but I cannot get it.
> To achieve the goal--notification handler deletion--the `~WalletModel` destructor must be called. But I cannot see the way how deletion of the `m_wallet_controller` could works here, as an instance of the `WalletController` class does _not_ own `WalletModel` objects, because `WalletController::m_wallets` is `std::vector<WalletModel*>`, not `std::vector<std::unique_ptr<WalletModel>>`.
> 
> Did I miss the point?

No, these is a reasonable question because wallet controller's ownership of wallet models is implicit instead of explicit. But `m_wallet` vector usage is just emulating `unique_ptr` without using `unique_ptr` for some reason (probably no reason).

In the case where an `m_wallet` entry is erased the pointer is deleted:

https://github.com/bitcoin-core/gui/blob/d67330d11245b11fbdd5e2dd5343ee451186931e/src/qt/walletcontroller.cpp#L179-L186

In the case where the `m_wallet` vector is cleared, the pointers are deleted:

https://github.com/bitcoin-core/gui/blob/d67330d11245b11fbdd5e2dd5343ee451186931e/src/qt/walletcontroller.cpp#L108-L114

And in the case where the controller object is destroyed, any models that weren't cleared or erased are destroyed:

https://github.com/bitcoin-core/gui/blob/d67330d11245b11fbdd5e2dd5343ee451186931e/src/qt/walletcontroller.cpp#L142-L146

I could add more details to this comment, but I don't think it would be appropriate to mention implementation details of the wallet controller outside the wallet controller (especially when implementation seems unnecessarily complicated and could probably be simplified later). So hopefully that clears things up and things for looking into this!